logr.go

```diff
@@ -180,3 +184,20 @@ type Logger interface {
 // InfoLogger provides compatibility with code that relies on the v0.1.0 interface
 // Deprecated: use Logger instead. This will be removed in a future release.
 type InfoLogger = Logger
+
+type contextKey struct{}
+
+// FromContext returns a logr.Logger constructed from the context or returns a null logger
+// if no no logger details are found in the context.
+func FromContext(ctx context.Context) Logger {
+	if v, ok := ctx.Value(contextKey{}).(Logger); ok {
+		return v
+	}
+
+	return nullLogger{}
```

Comment:
From #25:

I think the contract here should be "returns the context's Logger or nil if no Logger is in the context". Libs which expect a Logger should say so and make callers provide one. I know DI is not really Go-ish, but it has saved my butt many times.

Offering a NullLogger (#26) and automatically using it are 2 very different things.

Just return nil here and make the comment clear.

Reply:
I think a disabled logger is a better experience. Then there is no difference for library code between getting a logger that has been disabled manually due to V or not having a logger in the context.  Otherwise we force everyone to perform a nil check every time they obtain a logger from a context.

Reply:
Github email put this on the conversation rather than here.

I understand that you think that.  We disagree on this point and that is
OK.  I'm a fan of low-magic and obvious errors.

If it turns out I am wrong (it happens at least daily, ask my kids!) this
can probably be changed with relative ease.

Reply:
As mentioned in #25: we could add simple FromContext() and FromContextOrNull() which does what you describe, I guess.

Reply:
I'll implement whatever the agreed consensus is on #25

Reply:
#26 is merged and I think we're at consensus on #25 ?